### ISSUE TITLE
Cannot reproduce the results

### ISSUE THREAD

**Issue by zhuqunxi** [AlexandreNap/Denoising-Diffusion-Probabilistic-Model-toy-project#1]
Hi,

This project is beneficial for me as a beginner of the DDPM. However, I cannot reproduce the result without any changes. Could you provide the requirements so that I can match the packages used in the project? Thanks.

<img width="411" alt="ae7d2ec2740939b95d6127cc7677a7b" src="https://github.com/AlexandreNap/Denoising-Diffusion-Probabilistic-Model-toy-project/assets/22273557/a94f4767-62d1-4267-8617-e4dec3d33c4e">

Figure 1. Cannot generate the circle image using the original code.

<img width="604" alt="7cdeda035bbd74fb1cfd7b416b04a05" src="https://github.com/AlexandreNap/Denoising-Diffusion-Probabilistic-Model-toy-project/assets/22273557/5baf62c3-c2ef-4d31-aeea-f8283c8589bb">

Figure 2. Therefore, I produce a new circle in this way.

<img width="510" alt="66ba77e3523cefb8682ccf1338d9959" src="https://github.com/AlexandreNap/Denoising-Diffusion-Probabilistic-Model-toy-project/assets/22273557/89d04f72-6b59-46c5-b2da-af838cdd75ae">

Figure 3. Using the above new circle as data, the results are bad.

Best,

Qunxi

**Comment by AlexandreNap** [comment]
Hi,

I ran my notebook with a recent env and got the same issues as you.
On your first point, the dataset creation was broken because of some changes in PILImage.convert('p') converting pixels in the range [0,225] instead of expected [0,255]. There was then no null probabilities of generating a datapoint uniformly everywhere in the image. Now fixed in recent commits.

On your second point, in train() function i left the dataloader without shuffle since i was using a random generated dataset. But with your dataset shuffling was essential. After adding it and trying differents learning rate, I could have nice results :

![Capture d'écran 2023-10-27 181446](https://github.com/AlexandreNap/Denoising-Diffusion-Probabilistic-Model-toy-project/assets/85485132/e8bc79fa-9927-4331-a7f9-b981355bab9d)

![Capture d'écran 2023-10-27 181646](https://github.com/AlexandreNap/Denoising-Diffusion-Probabilistic-Model-toy-project/assets/85485132/1ea0e060-2298-45f9-9f05-abc8cd409795)

Please try updated repo, it should be alright now.

**Comment by zhuqunxi** [comment]
> Hi,
> 
> I ran my notebook with a recent env and got the same issues as you. On your first point, the dataset creation was broken because of some changes in PILImage.convert('p') converting pixels in the range [0,225] instead of expected [0,255]. There was then no null probabilities of generating a datapoint uniformly everywhere in the image. Now fixed in recent commits.
> 
> On your second point, in train() function i left the dataloader without shuffle since i was using a random generated dataset. But with your dataset shuffling was essential. After adding it and trying differents learning rate, I could have nice results :
> 
> ![Capture d'écran 2023-10-27 181446](https://user-images.githubusercontent.com/85485132/278709430-e8bc79fa-9927-4331-a7f9-b981355bab9d.png)
> 
> ![Capture d'écran 2023-10-27 181646](https://user-images.githubusercontent.com/85485132/278709541-1ea0e060-2298-45f9-9f05-abc8cd409795.png)
> 
> Please try updated repo, it should be alright now.

Thank you very much! My concerns are all solved.
